### ISSUE TITLE
Build Gambit installers on CI pipeline

### ISSUE THREAD

**Issue by daviwil** [gambit/gambit#594]
This issue is meant to track the work to get Gambit installers building on a GitHub Actions release pipeline.  This might be used to automate releases when new tags are pushed to the repo.

**Comment by daviwil** [comment]
@feeley - I've started setting up a release pipeline but I'm hitting a `*** FATAL ERROR - Heap overflow` when using the bootstrapped `gsc-boot.exe` to build the latest `gsc`.  I decided to bump up the heap size when running `gsc-boot` (`-:h200000`) and I'm still getting [errors when building `_gsc.scm`](https://github.com/daviwil/gambit/runs/754717370?check_suite_focus=true#step:4:1953):

```
../gsc-boot -:h2000000,~~bin=../bin,~~lib=../lib,~~include=../include -f  -prelude "(##namespace(\"c#\"))(##include\"../lib/header.scm\")" -warnings -c _gsc.scm
/bin/sh: ../gsc-boot: Resource temporarily unavailable
```

Is it possible that we're running out of memory on the build machine?  Is there any other parameter I could tweak to help alleviate this issue?

**Comment by daviwil** [comment]
I'm trying to put together the optimized build based on `make dist` but I find myself a little uncertain about the process.  Does this sound right?

**CI build on `master`**

1. Bootstrap Gambit from scratch
2. `make dist` in the _root_ of the repo to gather the source of the bootstrapped `gsc`
3. Store the resulting `tgz` as an artifact for that OS/arch.

**PR build**

1. Clone the repo without tags so that the dummy `gsc-boot` will be used
2. Download the source `tgz` from the last successful build on `master`
3. Unpack `.c` files from the `tgz` into `gsc/` and `lib/` into the repo folder
4. Run `make`, etc to build everything as usual

Should I be packing up `make dist` from `boot/` instead and then unpack it to `boot/` in the PR build, build it, `make bootstrap` and then run `make` on the repo folder?

**Comment by daviwil** [comment]
btw @feeley, something about one of the latest commits to `master` (https://github.com/gambit/gambit/commit/5bd523790d4d548b6a8cef4be53fa80a51d18bf6) has broken tests on MinGW builds: 

https://github.com/gambit/gambit/runs/758606784#step:5:170

**Comment by feeley** [comment]
> I'm trying to put together the optimized build based on `make dist` but I find myself a little uncertain about the process. Does this sound right?
> 
> **CI build on `master`**
> 
> 1. Bootstrap Gambit from scratch
> 2. `make dist` in the _root_ of the repo to gather the source of the bootstrapped `gsc`
> 3. Store the resulting `tgz` as an artifact for that OS/arch.

Yes.  Except the OS/arch have *no* impact on the `tgz`.  That is one of the beauties of Gambit... the generated C code is OS and architecture independent.  So you only need one `tgz` artifact.

> **PR build**
> 
> 1. Clone the repo without tags so that the dummy `gsc-boot` will be used
> 2. Download the source `tgz` from the last successful build on `master` with the same OS/arch
> 3. Unpack `.c` files from the `tgz` into `gsc/`, `gsi/`, and `lib/` into the repo folder
> 4. Run `make`, etc to build everything as usual

No.  The dummy `gsc-boot` is only usable for building a `make dist` snapshot (which includes releases).  So at step 3 you should unpack the `tgz` and rename it `boot`, then do a `make` in `boot`, which will create a `boot/gsc/gsc` that is consistent with the last successful build on `master`.  Then I think a `make` in Gambit's root will succeed... it should build another `boot/gsc/gsc` using the sources in `gsc`, then use that new `gsc` to compile the .scm files.

This reminds me that the version test in the `boot-wip-compiler` target will probably fail at the next release because the version information now includes the git commit hash.  This needs to be fixed.

**Comment by feeley** [comment]
> btw @feeley, something about one of the latest commits to `master` ([5bd5237](https://github.com/gambit/gambit/commit/5bd523790d4d548b6a8cef4be53fa80a51d18bf6)) has broken tests on MinGW builds:
> 
> https://github.com/gambit/gambit/runs/758606784#step:5:170

Can you reproduce this?  I have noticed random failures on Windows in the past and never determined what was the cause.

**Comment by daviwil** [comment]
> So at step 3 you should unpack the `tgz` and rename it `boot`, then do a `make` in `boot`, which will create a `boot/gsc/gsc` that is consistent with the last successful build on `master`. Then I think a `make` in Gambit's root will succeed... it should build another `boot/gsc/gsc` using the sources in `gsc`, then use that new `gsc` to compile the .scm files.

OK, I'll give that a shot!

I've only seen the test issue on CI runs after that commit, I'll try to reproduce it locally.

**Comment by daviwil** [comment]
Hmmm, perhaps it's not consistent, I saw your latest PR build succeeded.

**Comment by feeley** [comment]
So with ec8d10d1af3fc763f1ab20b962198beee50173d7 the build time of `make -j8` is now 25% faster (on my machine it saves 10 seconds on what used to be a 40 second build time).  For the CI, it would be good to automatically detect the number of cores and then use that N in a `make -jN`.  On macOS and linux these work:

    nproc  # linux

    sysctl -n hw.physicalcpu  # macOS

    echo %NUMBER_OF_PROCESSORS%  # Windows

**Comment by daviwil** [comment]
Cool!  I'll give that a shot.

**Comment by feeley** [comment]
The build times on Windows are about 40 minutes, or 2x the build time on linux and macOS.  So we have to work on making the Windows builds faster... a lot faster.  I think they are creating artifacts that aren't created for linux and macOS, so maybe that's the first thing to remove.  Also, I wonder if the C compilation time could be improved by removing the `--enable-single-host`.  In fact, not using that flag may improve the diversity of the testing because currently all builds use that flag.  Anyway, that's something to experiment with.

**Comment by daviwil** [comment]
I'm considering making a different workflow entirely for PR builds that eliminates the "usable" Gambit builds so that the main `ci.yml` doesn't get too complicated with conditional checks.  What's the bare minimum you'd like to see there for build/test scenarios?

**Comment by feeley** [comment]
The tests of the Windows builds are failing again... I suspect it is the `open-process` of the test runner that gets interrupted.  If you have access to a Windows machine, can you test this?  Check if a manual `make ut` works.

**Comment by daviwil** [comment]
Yep, I'm in one right now, will run it and check.

**Comment by feeley** [comment]
> I'm considering making a different workflow entirely for PR builds (so that the main `ci.yml` doesn't get too complicated with conditional checks) which eliminates the "usable" Gambit builds and anything else that wastes time. What's the bare minimum you'd like to see there for build/test scenarios?

The bare minimum is building Gambit (bootstrapping from the result of the most recent successful `make dist`) and passing `make modules` and `make check`.  Basically the steps that are in `README.md`.

**Comment by lassik** [comment]
Could we add `make boot.tgz` to create a `boot.tgz` such that `tar -xf boot.tgz` extracts into the `boot` directory?

It can start out by making the same tar file as `make dist` (except the name of the directory where it extracts would be `boot` instead of `gambit-v1_2_3`). If we figure out that it can work with fewer files, we can adjust the make target later.

**Comment by feeley** [comment]
The `make check` takes ~22 seconds currently because all tests are done sequentially.  I think this could be brought down to ~8 seconds (the duration of the longest test) by a fairly simple parallelization of the tests (when `-jN` is used).  I'll look into this.  Perhaps we can have the whole basic CI cycle take 90 seconds... that would be awesome!

**Comment by daviwil** [comment]
CI cycle on Windows will still require an additional 1.5-2 minutes for MSYS2 setup, but that's still not too bad!

**Comment by feeley** [comment]
@antoine-doucet is working on self-hosting the CI... perhaps we could have a machine where MSYS2 is already setup and ready to go...

**Comment by daviwil** [comment]
Couldn't reproduce the `make ut` issue on my local mingw64 installation, I wonder if PR #601 will fix it.

**Comment by antoine-doucet** [comment]
Really easy to set-up a self-hosted runner, I'll investigate if docker could be a fast and simple solution for the windows build

**Comment by daviwil** [comment]
The one concern I have about self-hosted runners is that if we start depending on pre-installed software on your runners, I won't be able to test CI changes on my fork.  It'd be ideal if the same CI configuration runs on both the hosted runners and the public GitHub images, otherwise it'll be very difficult for outside contributors to help improve it.

**Comment by feeley** [comment]
I downloaded the MINGW64 artifact to a Windows 10 machine to test some things.  When I ran gsi.exe I got this error:

    The code execution cannot proceed because libwinpthread-1.dll was not found.
    Reinstalling the program may fix this problem.

This should not happen... the gsi.exe executable should be self contained.  There must be some link flag that bundles that dll into gsi.exe .

**Comment by daviwil** [comment]
I'll figure out how to get that statically linked, found some clues on StackOverflow already.

**Comment by daviwil** [comment]
@feeley It seems that merely adding `-static` to the end of this line is enough to fix that issue:

https://github.com/gambit/gambit/blob/master/gsi/makefile.in#L169

Since we probably don't want that flag used on every platform, what's the correct way to make sure it only gets applied in MinGW builds?  Is there some way I can funnel it through from `configure`?

**Comment by feeley** [comment]
`./configure CC="gcc -static"` or set `CFLAGS`.

**Comment by daviwil** [comment]
`CC="gcc -static"` worked well, I added it to PR #601.

**Comment by daviwil** [comment]
I'm almost finished creating a separate CI pipeline for pull requests to shorten its duration but I'm investigating some final issues with it.  Hopefully I'll have a PR out in the next couple of days!
